symengine/cwrapper.cpp

```diff
@@ -200,4 +210,78 @@ int is_a_Symbol(const basic c)
     return is_a<Symbol>(*(*RCP_const_cast(c)));
 }
 
+
+// C wrapper for std::vector<int>
+
+struct CVectorInt {
+    std::vector<int> m;
+};
+
+CVectorInt* vectorint_new()
+{
+    return new CVectorInt;
+}
+
+int vectorint_placement_new(CVectorInt *self, size_t size)
+{
+    if (!SymEngine::is_aligned(self)) return 2;
+    if (size < sizeof(CVectorInt)) return 1;
+    new(self) CVectorInt;
+    return 0;
+}
+
+void vectorint_free(CVectorInt *self)
+{
+    delete self;
+}
```

Comment:
You should not call delete on it. The caller provided the memory, the
caller is responsible to free it. However, you are right that we need to
call the destructor ourselves, i.e. provide a function to do that. Perhaps
call it vectorint_placement_free or something like that, and it would just
call the destructor.

Sent from my mobile phone.
On Jul 1, 2015 6:37 AM, "Abinash Meher" <notifications@github.com> wrote:

> In symengine/cwrapper.cpp
> <https://github.com/sympy/symengine/pull/495#discussion_r33673666>:
>
> > +{
> > +    return new CVectorInt;
> > +}
> > +
> > +int vectorint_placement_new(CVectorInt *self, size_t size)
> > +{
> > +    if (!SymEngine::is_aligned(self)) return 2;
> > +    if (size < sizeof(CVectorInt)) return 1;
> > +    new(self) CVectorInt;
> > +    return 0;
> > +}
> > +
> > +void vectorint_free(CVectorInt *self)
> > +{
> > +    delete self;
> > +}
>
> Let's say we have more memory than sizeof(CVectorInt) allocated on the
> stack, and we use vector_placement_new to initialize CVectorInt at the
> given location. So, when we call delete on the pointer to that location,
> will the whole memory(that is more than the size of CVectorInt) be freed?
> If I am guessing correctly, does SymEngine::is_aligned has anything to do
> to ensure that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/symengine/pull/495/files#r33673666>.
>